### PR TITLE
Fix duplicate messages created by jobs>1.

### LINKED ISSUE #3584: Same error is reported multiple times when using multiprocessing (including performance decrease)
### Steps to reproduce
1. create a new virtualenv with latest pylint development state (3fffdf03d5a32e4a0d1b5216ea9d16dacacdaa6f)
2. create a folder "test" somewhere
3. add an (empty) "__init__.py" file there
4. add an (empty "spam1.py" file there, write "asdf" and nothing else in that file
5. copy the file 20 times in this same folder, calling the files "spam2.py" etc.
6. go to the parent directory of "test" and call "pylint -j 1 -E test":
```
(pylint) C:\work\pylint_test>pylint -j 1 -E test
************* Module test.spam1
test\spam1.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam10
test\spam10.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam11
test\spam11.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam12
test\spam12.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam13
test\spam13.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam14
test\spam14.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam15
test\spam15.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam16
test\spam16.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam17
test\spam17.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam18
test\spam18.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam19
test\spam19.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam2
test\spam2.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam20
test\spam20.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam3
test\spam3.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam4
test\spam4.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam5
test\spam5.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam6
test\spam6.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam7
test\spam7.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam8
test\spam8.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam9
test\spam9.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
```
7. go to the parent directory of "test" and call "pylint -j 0 -E test":
```
(pylint) C:\work\pylint_test>pylint -j 0 -E test
************* Module test.spam11
test\spam11.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam14
test\spam14.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam12
test\spam12.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
test\spam11.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam17
test\spam17.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam10
test\spam10.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam1
test\spam1.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
test\spam14.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam18
test\spam18.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam13
test\spam13.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam16
test\spam16.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam15
test\spam15.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
test\spam11.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
test\spam17.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam2
test\spam2.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
test\spam12.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam19
test\spam19.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
test\spam10.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam20
test\spam20.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
test\spam14.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
test\spam18.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam4
test\spam4.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
test\spam1.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam3
test\spam3.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
test\spam13.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam5
test\spam5.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
test\spam15.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam7
test\spam7.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
test\spam11.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
test\spam17.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
test\spam2.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam8
test\spam8.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
test\spam12.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
test\spam19.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam9
test\spam9.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
test\spam16.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam6
test\spam6.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
```
For example `test\spam11.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)` is reported four times.

### Current behavior
The same error is reported multiple times.

### Expected behavior
Each error is reported only once (and in the report at the correct place).

### pylint --version output
```
pylint 2.6.0-dev1
astroid 2.4.0
Python 3.6.4 (v3.6.4:d48eceb, Dec 19 2017, 06:54:40) [MSC v.1900 64 bit (AMD64)]
```

### PR DESCRIPTION
**Pull request by moser**
## Steps

- [X] Add yourself to CONTRIBUTORS if you are a new contributor.
- [X] Add a ChangeLog entry describing what your PR does.
- [x] If it's a new feature or an important bug fix, add a What's New entry in `doc/whatsnew/<current release.rst>`.
- [X] Write a good description on what the PR does.

## Description
Fixed an issue in `check_parallel` that re-emits all messages a process has seen before for each file. Added a `reset` method to the `CollectingReporter` and call it after each file a process analyses.

## Type of Changes

|   | Type |
| ------------- | ------------- |
| ✓  | :bug: Bug fix  |

## Related Issue
Closes #3584

### CODE REVIEW

**Comment by turbaszek** [comment]
When can we expect this to be released?

**Comment by revmischa** [comment]
Would love to see this released!

File: tests/lint/unittest_lint.py
Comment (Pierre-Sassoulas) [inline tests/lint/unittest_lint.py:794]: Unless I'm mistaken we can use `REGRTEST_DATA_DIR` directly ?

**Review by Pierre-Sassoulas — APPROVED** [review]
Looks good, thank you for the fix. I made a minor comment let me know what you think.

File: tests/lint/unittest_lint.py
Comment (moser) [reply tests/lint/unittest_lint.py]: Yeah, that was a left-over from the evolution of the test :-)
Fixed.

**Comment by moser** [comment]
@Pierre-Sassoulas 
Will you "add a What's New entry in doc/whatsnew/<current release.rst>" or do I need to do it before you can merge the PR?

**Comment by Pierre-Sassoulas** [comment]
Oops you're right, I forgot about whatsnew, and it's a notable fix. Could you add something, please ? It's better if the merge commit is atomic so we can revert in case something goes wrong.

**Review by Apakottur — APPROVED** [review]
Would be happy to see this merged.

**Comment by moser** [comment]
@Pierre-Sassoulas I added the whatsnew entry

**Comment by Pierre-Sassoulas** [comment]
Thank you, great fix !

**Comment by turbaszek** [comment]
Thanks @moser for the fix 🚀 @Pierre-Sassoulas when can we expect this fix to be released?

**Comment by Pierre-Sassoulas** [comment]
Probably in 2.6.2 or right now with `python -m pip install git+https://github.com/PyCQA/pylint/commit/da73eb77b459f77ba3ca7ce1fd5632da40e3ae14#egg=pylint` :)

**Comment by aleksandr-kotlyar** [comment]
> Probably in 2.6.2 or right now with `python -m pip install git+https://github.com/PyCQA/pylint/commit/da73eb77b459f77ba3ca7ce1fd5632da40e3ae14#egg=pylint` :)

Can't wait for release 💓 

x8 logs are annoying.

**Comment by DavidCain** [comment]
For anybody else excited to see this released soon, looks like it didn't actually make it out for 2.6.2, but looks like it's slated for 2.7.0

----

> [2.7 will be released] very soon (once we make sure there is no problem with astroid 2.5 by checking the issues that were reported today).

 https://github.com/PyCQA/pylint/issues/4099#issuecomment-780086037
 
 (You can see that it's slated to go out with 2.7.0): https://github.com/PyCQA/pylint/commit/4faccd0563c628954f04322ffb696905a3dab62d
 
 Thanks for all you do, Pierre, and thanks for this patch, moser! :smile:

**Pull request merged**
